### PR TITLE
#640: Add BITFIELD_RO command

### PR DESCRIPTION
**Pull request by iamskp11**
This PR closes [#640](https://github.com/DiceDB/dice/issues/640) - BITFIELD_RO command.

1. Integration tests
image

2. Unit tests
image

### CODE REVIEW

**Comment by iamskp11** [comment]
Hey @JyotinderSingh @lucifercr07 @apoorvyadav1111 , please review this PR for adding BITFIELD_RO command. 
Thanks

**Comment by iamskp11** [comment]
> Hi @iamskp11 , Thanks for the changes. I have some suggestions regarding refactoring the code for both evals to ensure consistency in any future modifications in one. I ll be reviewing the code thoroughly in the morning and comment the suggestions, if you would consider them. Thanks

Sure @apoorvyadav1111 . 
Also, I went through Redis implementation for BITFIELD and BITFIELD_RO, and they have simply used a flag, to differentiate between the two. I was thinking maybe we can also do similar.

**Comment by apoorvyadav1111** [comment]
Yes I had similar in mind. Although we can have two functions; one to parse ops and another to run ops. The one which parses ops need the flag. The one which runs it will not will execute the ops irrespective of the type of command .

**Comment by iamskp11** [comment]
> Yes I had similar in mind. Although we can have two functions; one to parse ops and another to run ops. The one which parses ops need the flag. The one which runs it will not will execute the ops irrespective of the type of command .

Didn't get this @apoorvyadav1111 . I think parsing ops function should not worry about command/flag. Let that parse the ops, and if we get update/set commands, we can throw error if it is with RO flag.

**Comment by iamskp11** [comment]
Hey @apoorvyadav1111  @JyotinderSingh @lucifercr07 , can you have a look at https://github.com/DiceDB/dice/pull/994 , and review it. It uses a generic method for BITFIELD and BITFIELD_RO commands, with a flag for allowing only GET subcommand.

**Comment by apoorvyadav1111** [comment]
Hi @iamskp11 , the idea is to change the current bitfield and bitfield_ro commands into :
1. `parseOps(args, isReadOnly) (ops BitFieldOp[], err)` : function returns ops array or error if parseing failed.
2. `executeOps(ops)`: function runs the operations and internally returns the result or error.

With these changes, pseudocode for bitfield would be:
```
ops:= parseOps(args, false)
return executeOps(ops)
```
and bitfield_ro would be:
```
ops:= parseOps(args, true)
return executeOps(ops)
```

You can make changes according to when you refactor them. The idea is to reuse the logic in bitfield.
If need be, I can make code suggestions in the review as well.

Apoorv

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]
Hi, let's improve the bitfield and bitfield_ro as discussed. We can connect over discord to discuss more.

**Comment by iamskp11** [comment]
@apoorvyadav1111 , thats great. Now, I got your idea. Will share the PR soon. Thanks a lot!

**Comment by iamskp11** [comment]
Hey @apoorvyadav1111 , have done the changes as per your suggestion. Integration and unit tests work fine for both BITFIELD and BITFIELD_RO. Please review.
Requesting @JyotinderSingh @lucifercr07 to have a look and review.

**Review by apoorvyadav1111 — APPROVED** [review]
Hi, the code looks good to me and is what I had in mind. Regarding moving parsing functions to utils, I am not sure. Maybe @JyotinderSingh can share his views on this. In any case, it should be a small change even if we move the code back to eval.go.

Thanks for taking up my suggestion.
Apoorv

File: internal/server/utils/bitfield.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	Kind   string
+	EType  string
+	EVal   int64
+	Offset int64
+	Value  int64
+}
+
+// parseEncodingAndOffet function parses offset and encoding type for bitfield commands
+// as this part is common to all subcommands
+func parseEncodingAndOffset(args []string) (eType, eVal, offset interface{}, err error) {
```
Comment (JyotinderSingh) [inline internal/server/utils/bitfield.go:20]: Maybe we can name this better
```suggestion
func parseBitfieldEncodingAndOffset(args []string) (eType, eVal, offset interface{}, err error) {
```

File: internal/eval/eval.go [hunk 2]
```diff
@@ -4816,165 +4816,43 @@ func evalZRANGE(args []string, store *dstore.Store) []byte {
...
-			return diceerrors.NewErrWithMessage(diceerrors.SyntaxErr)
+			overflowType = op.EType
 		}
 	}
+	return result
+}
+
+// Generic method for both BITFIELD and BITFIELD_RO.
+// isReadOnly method is true for BITFIELD_RO command.
+func evalBITFIELDGeneric(args []string, store *dstore.Store, isReadOnly bool) []byte {
```
Comment (JyotinderSingh) [inline internal/eval/eval.go:4848]: Let's reserve the eval prefix for top level eval commands. We can rename this as follows.
```suggestion
func bitfieldEvalHelper(args []string, store *dstore.Store, isReadOnly bool) []byte {
```

File: internal/eval/eval.go [hunk 3]
```diff
@@ -4816,165 +4816,43 @@ func evalZRANGE(args []string, store *dstore.Store) []byte {
...
-		EVal   int64
-		Offset int64
-		Value  int64
-	}
-	var ops []BitFieldOp
-
-	for i := 1; i < len(args); {
-		switch strings.ToUpper(args[i]) {
+// This method executes each operation, contained in ops array, based on commands used.
+func executeOps(value *ByteArray, ops []utils.BitFieldOp) []interface{} {
```
Comment (JyotinderSingh) [inline internal/eval/eval.go:4820]: It's unclear what the executeOps method does by just looking at its naming.

```suggestion
func executeBitfieldOps(value *ByteArray, ops []utils.BitFieldOp) []interface{} {
```

File: internal/eval/eval.go [hunk 4]
```diff
@@ -4816,165 +4816,43 @@ func evalZRANGE(args []string, store *dstore.Store) []byte {
...
 		}
 	}
+	return result
+}
+
+// Generic method for both BITFIELD and BITFIELD_RO.
+// isReadOnly method is true for BITFIELD_RO command.
+func evalBITFIELDGeneric(args []string, store *dstore.Store, isReadOnly bool) []byte {
+	var ops []utils.BitFieldOp
+	ops, err2 := utils.ParseOps(args, isReadOnly)
```
Comment (JyotinderSingh) [inline internal/eval/eval.go:4850]:
```suggestion
	ops, err2 := utils.ParseBitfieldOps(args, isReadOnly)
```

**Review by JyotinderSingh — CHANGES_REQUESTED** [review]

**Comment by iamskp11** [comment]
Have done the function name changes , @JyotinderSingh . PR is now ready for review

**Comment by apoorvyadav1111** [comment]
Hi @iamskp11 , Please respond and resolve the review comments from @JyotinderSingh .

**Comment by iamskp11** [comment]
> Hi @iamskp11 , Please respond and resolve the review comments from @JyotinderSingh .


Hi @apoorvyadav1111 , have done that in my previous commit. It's ready for review.

File: internal/server/utils/bitfield.go (hunk 1 above)
Comment (iamskp11) [reply internal/server/utils/bitfield.go]: Updated in this [commit](https://github.com/DiceDB/dice/commit/b4601acfe13a6de265db09477dc84a3597e2fe26)

File: internal/eval/eval.go (hunk 3 above)
Comment (iamskp11) [reply internal/eval/eval.go]: Updated to **executeBitfieldOps** in this [commit](https://github.com/DiceDB/dice/commit/b4601acfe13a6de265db09477dc84a3597e2fe26)

File: internal/eval/eval.go (hunk 4 above)
Comment (iamskp11) [reply internal/eval/eval.go]: Changed to **ParseBitfieldOps** in this [commit](https://github.com/DiceDB/dice/commit/b4601acfe13a6de265db09477dc84a3597e2fe26)

**Review by JyotinderSingh — APPROVED** [review]

**Pull request merged**